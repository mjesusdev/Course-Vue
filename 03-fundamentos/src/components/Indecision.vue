<template>
  <div>
    <img v-if="img" :src="img" alt="bg">
    <div class="bg-dark"></div>

    <div class="indecision-container">
      <input 
        type="text" 
        v-model="question"
        placeholder="Hazme una pregunta"
      >
      <p>Recuerda terminar con un signo de interrogación (?)</p>

      <div v-if="isValidQuestion">
        <h2>{{ question }}</h2>
        <h1>{{ answer }}</h1>
      </div>
    </div>
  </div>
</template>

<script>
export default {
    data() {
      return {
        question: null,
        answer: null,
        img: null,
        isValidQuestion: false
      }
    },
    methods: {
      async getAnswer() {
        this.answer = 'Pensando...'
        const { answer, image } = await fetch('https://yesno.wtf/api').then(res => res.json() )
      
        this.answer = answer === 'yes' ? 'Sí!' : 'No!'
        this.img = image
      }
    },
    watch: {
      question( value ) {

        this.isValidQuestion = false

        console.log({ value });

        if ( !value.includes('?') ) return

        this.isValidQuestion = true

        // TODO: Realizar la petición a la API
        this.getAnswer()
      }
    }
}
</script>

<style scoped>

    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 2%;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }

</style>