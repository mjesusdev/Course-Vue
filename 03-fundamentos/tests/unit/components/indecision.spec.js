import { shallowMount } from '@vue/test-utils'
import Indecision from '@/components/Indecision';

describe('Component Indecision', () => {

    let wrapper
    let clgSpy

    global.fetch = jest.fn( () => Promise.resolve({
        json: () => Promise.resolve({
            answer: 'yes',
            forced: false,
            image: 'https://yesno.wtf/assets/yes/2.gif'
        })
    }) )

    beforeEach(() => {
        wrapper = shallowMount( Indecision )

        clgSpy = jest.spyOn(console, 'log')

        jest.clearAllMocks()
    })

    test('Should do match with the snapshot', () => {
        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('Write in the input not should nothing', async() => {
        const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer' )

        const input = wrapper.find('input')
        await input.setValue('Hello World!')

        expect( clgSpy ).toHaveBeenCalledTimes(1)
        expect( getAnswerSpy ).not.toHaveBeenCalled() 
    })
    
    test('Write "?" should return the getAnswer', async() => {
        const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer' )
    
        const input = wrapper.find('input')
        await input.setValue('Hello World?')
    
        expect( clgSpy ).toHaveBeenCalledTimes(1)
        expect( getAnswerSpy ).toHaveBeenCalled()
    })

    test('Test in getAnswer', async() => {
        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect( img.exists() ).toBeTruthy()
        expect( wrapper.vm.img ).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect( wrapper.vm.answer ).toBe('SÃ­!')
    })

    test('Test in getAnswer - Error in API', async() => {
        fetch.mockImplementationOnce( () => Promise.reject('API is down') )

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect( img.exists() ).toBeFalsy()
        expect( wrapper.vm.answer ).toBe('The API cannot start')
    })

})