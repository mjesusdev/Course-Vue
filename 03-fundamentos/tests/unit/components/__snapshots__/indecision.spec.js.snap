// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Indecision Should do match with the snapshot 1`] = `
<div>
  <!--v-if-->
  <div class="bg-dark"></div>
  <div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
    <p>Recuerda terminar con un signo de interrogación (?)</p>
    <!--v-if-->
  </div>
</div>
`;

exports[`Component Indecision should do match with the snapshot 1`] = `
<div>
  <!--v-if-->
  <div class="bg-dark"></div>
  <div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
    <p>Recuerda terminar con un signo de interrogación (?)</p>
    <!--v-if-->
  </div>
</div>
`;
